File: pac-man-board-game/Game/Items/Player.cs
namespace pacMan.Game.Items;

public interface IPlayer
{
    string Name { get; init; }
    Character PacMan { get; init; }
    string Colour { get; init; }
    Box? Box { get; init; }
    State State { get; set; }
}

public enum State
{
    WaitingForPlayers,
    Ready,
    InGame
}

public class Player : IPlayer, IEquatable<Player>
{
    public bool Equals(Player? other)
    {
        if (ReferenceEquals(null, other)) return false;
        if (ReferenceEquals(this, other)) return true;
        return Name == other.Name && PacMan.Equals(other.PacMan) && Colour == other.Colour && Box.Equals(other.Box) &&
               State == other.State;
    }

    public required string Name { get; init; }
    public required Character PacMan { get; init; }
    public required string Colour { get; init; }
    public Box? Box { get; init; }
    public State State { get; set; } = State.WaitingForPlayers;

    public override bool Equals(object? obj)
    {
        if (ReferenceEquals(null, obj)) return false;
        if (ReferenceEquals(this, obj)) return true;
        return obj.GetType() == GetType() && Equals((Player)obj);
    }

    public override int GetHashCode() => HashCode.Combine(Name, PacMan, Colour, Box, (int)State);
}

File: pac-man-board-game/ClientApp/src/game/player.ts
import {Character, CharacterType} from "./character";
import Box from "./box";
import {getDefaultStore} from "jotai";
import {currentPlayerNameAtom, playersAtom} from "../utils/state";
import Pellet from "./pellet";
import rules from "./rules";

export enum State {
  waitingForPlayers,
  ready,
  inGame,
  disconnected
}

export default class Player implements PlayerProps {
  private static store = getDefaultStore();
  public readonly username;
  public readonly pacMan;
  public readonly colour;
  public readonly box;
  public state;

  constructor(props: PlayerProps) {
    this.username = props.username;
    this.colour = props.colour;
    this.box = new Box(props.box ?? {colour: props.colour});
    this.pacMan = new Character(props.pacMan ?? {
      colour: props.colour,
      type: CharacterType.pacMan
    });
    this.state = props.state ?? State.waitingForPlayers;
  }

  public isTurn(): boolean {
    return Player.store.get(currentPlayerNameAtom) === this.username;
  }

  public addPellet(pellet: Pellet): void {
    this.box.addPellet(pellet);
  }

  public stealFrom(other: Player): void {
    for (let i = 0; i < rules.maxStealPellets; i++) {
      const pellet = other.box.pellets.pop();
      if (pellet)
        this.box.addPellet(pellet);
    }
    Player.store.set(playersAtom, Player.store.get(playersAtom).map(player => player));
  }

}
